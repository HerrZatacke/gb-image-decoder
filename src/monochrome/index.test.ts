import { describe, expect, test } from 'vitest';
import { hash } from 'ohash';
import { getFullParams, getRawImageData } from '.';
import { MonochromeImageCreationParams, ExportFrameMode } from '../index';
import tiles1x1 from '../../test/data/tiles/monochrome/1x1';
import tiles16x14 from '../../test/data/tiles/monochrome/16x14';
import tiles20x18 from '../../test/data/tiles/monochrome/20x18';
import tiles20x18l1 from '../../test/data/tiles/monochrome/20x18_1';
import tiles20x86 from '../../test/data/tiles/monochrome/20x86';
import { bw, red } from '../../test/data/palettes';
import { writeImageFileFromImageData } from '../../test/helpers/writeImageFileFromImageData';

const tileSets: Record<string, string[]> = {
  '1x1': tiles1x1,
  '16x14': tiles16x14,
  '20x18': tiles20x18,
  '20x18_1': tiles20x18l1,
  '20x86': tiles20x86,
};

// const testScaleFactors = [1, 2];
const testScaleFactors = [2];
const testFrameModes: ExportFrameMode[] = [
  ExportFrameMode.FRAMEMODE_KEEP,
  ExportFrameMode.FRAMEMODE_CROP,
  ExportFrameMode.FRAMEMODE_SQUARE_BLACK,
  ExportFrameMode.FRAMEMODE_SQUARE_WHITE,
];

interface TestOption {
  tileKey: string,
  tilesPerLine: number,
  imageStartLine: number,
  scaleFactor: number,
  handleExportFrame: ExportFrameMode,
}

const options: TestOption[] = [
  { tileKey: '1x1', tilesPerLine: 1, imageStartLine: 0, scaleFactor: 1, handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP },
  { tileKey: '16x14', tilesPerLine: 16, imageStartLine: 2, scaleFactor: 1, handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP },
  { tileKey: '20x18_1', tilesPerLine: 20, imageStartLine: 1, scaleFactor: 1, handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP },
  { tileKey: '20x18', tilesPerLine: 20, imageStartLine: 2, scaleFactor: 1, handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP },
  { tileKey: '20x86', tilesPerLine: 20, imageStartLine: 13, scaleFactor: 1, handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP },
]
  .reduce((acc: TestOption[], option: TestOption): TestOption[] => ([
    ...acc,
    ...(
      testScaleFactors.map((scaleFactor) => ({
        ...option,
        scaleFactor,
      })).flat()
    ),
  ]), [])
  .reduce((acc: TestOption[], option: TestOption): TestOption[] => ([
    ...acc,
    ...(
      testFrameModes.map((handleExportFrame) => ({
        ...option,
        handleExportFrame,
      })).flat()
    ),
  ]), []);

describe('Monochrome image generation', () => {
  describe.each(options)('using $tileKey raw data scaled by $scaleFactor with frameMode $handleExportFrame', async ({ tileKey, tilesPerLine, imageStartLine, scaleFactor, handleExportFrame }) => {
    const decoderUpdateParams: MonochromeImageCreationParams = {
      framePalette: red,
      imagePalette: bw,
      tiles: tileSets[tileKey],
      imageStartLine,
      tilesPerLine,
      scaleFactor,
      handleExportFrame,
    };

    const fullParams = getFullParams(decoderUpdateParams);
    const {
      data: rawData,
      dimensions,
    } = getRawImageData(fullParams);


    test('generates expected hash for fullParams', () => {
      expect(hash(fullParams)).toMatchSnapshot();
    });

    test('generates expected dimensions', () => {
      expect(dimensions).toMatchSnapshot();
    });

    test('generates expected hash for raw image data', () => {
      expect(hash(rawData)).toMatchSnapshot();
    });

    test('writes image file without error', async () => {
      await writeImageFileFromImageData(
        `mono_${tileKey}_${scaleFactor}x_${handleExportFrame}_default.png`,
        rawData,
        dimensions,
      );
    });
  });
});
