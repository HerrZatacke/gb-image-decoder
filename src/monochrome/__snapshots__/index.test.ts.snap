// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'crop' > generates expected dimensions 1`] = `
{
  "height": 16,
  "width": 16,
}
`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'crop' > generates expected hash for fullParams 1`] = `"ZtepAnH9Hy"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'crop' > generates expected hash for raw image data 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'keep' > generates expected dimensions 1`] = `
{
  "height": 16,
  "width": 16,
}
`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'keep' > generates expected hash for fullParams 1`] = `"PfYOCQHfnl"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'keep' > generates expected hash for raw image data 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'square_black' > generates expected dimensions 1`] = `
{
  "height": 16,
  "width": 16,
}
`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for fullParams 1`] = `"UvLlwTs2IU"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for raw image data 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'square_white' > generates expected dimensions 1`] = `
{
  "height": 16,
  "width": 16,
}
`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for fullParams 1`] = `"gRIPeGNIx9"`;

exports[`Monochrome Decoder > using '1x1' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for raw image data 1`] = `"jwkXkBq86d"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'crop' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'crop' > generates expected hash for fullParams 1`] = `"6VfZOuypN6"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'crop' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'keep' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'keep' > generates expected hash for fullParams 1`] = `"Rp3ArTgkVf"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'keep' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'square_black' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for fullParams 1`] = `"Xk4QDeFOFf"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'square_white' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for fullParams 1`] = `"dRrA0mqOG6"`;

exports[`Monochrome Decoder > using '16x14' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'crop' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'crop' > generates expected dimensions 2`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'crop' > generates expected hash for fullParams 1`] = `"gdlV4jrxgH"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'crop' > generates expected hash for fullParams 2`] = `"gdlV4jrxgH"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'crop' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'crop' > generates expected hash for raw image data 2`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'keep' > generates expected dimensions 1`] = `
{
  "height": 288,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'keep' > generates expected dimensions 2`] = `
{
  "height": 288,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'keep' > generates expected hash for fullParams 1`] = `"RT7adivd4L"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'keep' > generates expected hash for fullParams 2`] = `"RT7adivd4L"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'keep' > generates expected hash for raw image data 1`] = `"dobrrcznjp"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'keep' > generates expected hash for raw image data 2`] = `"dobrrcznjp"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_black' > generates expected dimensions 1`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_black' > generates expected dimensions 2`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for fullParams 1`] = `"OjT6Qozyv7"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for fullParams 2`] = `"OjT6Qozyv7"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for raw image data 1`] = `"i3KxHIG14t"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for raw image data 2`] = `"i3KxHIG14t"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_white' > generates expected dimensions 1`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_white' > generates expected dimensions 2`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for fullParams 1`] = `"5gmp4NXLC2"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for fullParams 2`] = `"5gmp4NXLC2"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for raw image data 1`] = `"UJzaE6M2OG"`;

exports[`Monochrome Decoder > using '20x18' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for raw image data 2`] = `"UJzaE6M2OG"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'crop' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'crop' > generates expected hash for fullParams 1`] = `"hkk5nCxxe1"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'crop' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'keep' > generates expected dimensions 1`] = `
{
  "height": 288,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'keep' > generates expected hash for fullParams 1`] = `"CoY9A519lO"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'keep' > generates expected hash for raw image data 1`] = `"PotL6pdlaS"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'square_black' > generates expected dimensions 1`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for fullParams 1`] = `"R7WMWC2lZt"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for raw image data 1`] = `"dVXXCHqej0"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'square_white' > generates expected dimensions 1`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for fullParams 1`] = `"DSdsGxAnVF"`;

exports[`Monochrome Decoder > using '20x18_1' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for raw image data 1`] = `"qpDGmjQS58"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'crop' > generates expected dimensions 1`] = `
{
  "height": 224,
  "width": 256,
}
`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'crop' > generates expected hash for fullParams 1`] = `"nktZRWHN3C"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'crop' > generates expected hash for raw image data 1`] = `"vVfl7TJX14"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'keep' > generates expected dimensions 1`] = `
{
  "height": 688,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'keep' > generates expected hash for fullParams 1`] = `"ibz8vtUlhv"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'keep' > generates expected hash for raw image data 1`] = `"XICBvx6tL5"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'square_black' > generates expected dimensions 1`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for fullParams 1`] = `"sZSvrS7yVa"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'square_black' > generates expected hash for raw image data 1`] = `"p8bNTOi7Q0"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'square_white' > generates expected dimensions 1`] = `
{
  "height": 320,
  "width": 320,
}
`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for fullParams 1`] = `"BpJ88Fn9Ui"`;

exports[`Monochrome Decoder > using '20x86' raw data scaled by 2 with frameMode 'square_white' > generates expected hash for raw image data 1`] = `"p8bNTOi7Q0"`;
